### PR TITLE
Fix: Pasting a tag that is part of a transform and not matched ignores the content.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
Fixes: https://github.com/WordPress/gutenberg/issues/11057

On https://github.com/WordPress/gutenberg/issues/11057 @jrmd discovered that if we remove the preformatted block (wp.blocks.unregisterBlockType('core/preformatted')) when pasting some content inside the pre tag e.g `<pre>Pre</pre>`the content is simply ignored.

What happens is that we have other blocks that implement a transform for the PRE tag (the code block) but simple pre-content is not matched by the code block raw handler so in this case, we just returned nothing and ignored the content.

This PR changes the situation when a tag contains a transform that handles it but does not match its content we now create an HTML block to contain that content (filtered to make sure it is something the user can insert on the site).

## How has this been tested?
I removed the preformatted block `wp.blocks.unregisterBlockType('core/preformatted')`.
I pasted some HTML content inside a PRE tag e.g:`<pre>Pre</pre>` and checked it was added inside an HTML block.

### CODE REVIEW

**Comment by iseulde** [comment]
Could (integration) tests be added?

**Comment by jorgefilipecosta** [comment]
Hi @iseulde,
I was thinking of adding an end 2 end tests to check the case of removed preformatted I'm not certain if with integration we can test without preformatted being registered but I will have a look. Thank you for the suggestion.

File: packages/blocks/src/api/raw-handling/index.js [hunk 1]
```diff
@@ -190,7 +191,9 @@ export default function rawHandler( { HTML = '', plainText = '', mode = 'AUTO',
 					'Sanitized HTML: `' + node.outerHTML + '`'
 				);
 
-				return;
+				return createBlock( 'core/html', {
+					content: originalPieceFiltered,
```
Comment (noisysocks) [inline packages/blocks/src/api/raw-handling/index.js]: This isn't quite right. We will want to return the HTML of the current _node_ here, not the entire _piece_.

```js
return createBlock( 'core/html', {
	content: node.outerHTML,
} );
```

You can see this bug by:

1. Create a new post
2. Run `wp.blocks.unregisterBlockType( 'core/preformatted' )` in the console
3. Copy the **entire** text of this comment including the code snippet above
4. Paste it into Gutenberg

Only the code snippet above should appear as a HTML block. Instead, _the entire comment_ appears in the HTML block.

**Review by noisysocks — DISMISSED** [review]

File: packages/blocks/src/api/raw-handling/index.js (hunk 1 above)
Comment (noisysocks) [reply packages/blocks/src/api/raw-handling/index.js]: GIF of the above:

bug

**Comment by youknowriad** [comment]
Moving out of 4.3 as it doesn't seem urgent or ready.

**Comment by youknowriad** [comment]
I tested and confirmed that it works as advertised but I'm not sure I understand why we create a paragraph block instead of an HTML block like suggested initially?

**Comment by jorgefilipecosta** [comment]
Hi @youknowriad,

Our algorithm when we find some tag that we don't handle is to parse the children inside it as if the tag did not exist.
So right now (in master and this branch) if we paste "<form>aaaa</form>" we get a paragraph with "<p>aaaa</p>".
The pre tag is only accepted because two blocks (preformatted and code) set a handle for it. In master, if we remove both of them pasting "<pre>Pre</pre>" gives us "<p>Pre</p>".
What happened was that the code block accepted the pre tag but had a matcher to just accept some types of pre-content. We had a bug, where if a tag was accepted because of a specific raw handler transform, and its matcher did match the content, the content was discarded. In this PR we change that, the content now follows the general algorithm (parse the children inside it as if the tag did not exist).

The previous approach of returning HTML was not possible because it created a bug as @noisysocks discovered even if we are pasting content that can be handled (together with non matched content) our algorithm was not executed and we were just returning giant HTML block.

**Comment by youknowriad** [comment]
> The previous approach of returning HTML was not possible because it created a bug as @noisysocks discovered even if we are pasting content that can be handled (together with non matched content) our algorithm was not executed and we were just returning giant HTML block.

This feels like the algorithm should be reversed so that we consider this as a fallback or something but granted I don't know enough about it to say whether it's possible/good ...

I'll probably defer this review to @iseulde @mcsf

**Comment by catehstn** [comment]
@iseulde @mcsf Can you take a look at this PR when you have time, please?

**Comment by iseulde** [comment]
I think it needs a rebase after #12029. What issues are fixed here that that PR doesn't fix?

**Comment by jorgefilipecosta** [comment]
Hi @iseulde, this PR was rebased.
The core issue this PR address is still solved and the test case added here should fail on master.
The pre tag is only accepted on Gutenberg raw handling mechanism because two blocks preformatted and code registered raw transforms for it.
On master if we unregister the core block ```wp.blocks.unregisterBlockType('core/preformatted');```, and then we paste ```<pre>aaaaa</pre>``` we get an HTML block with an empty pre tag, the content was lost (bug).
If on master we continue the tests and also unregister code block ```wp.blocks.unregisterBlockType('core/code');```, we get a paragraph with aaaaa, this is right the content was not lost.

The bug is that if a given tag is only accepted because a block set a raw handler for it, but then is not matched on isMatch function we loose content.
On this branch, if a tag is accepted because of a rawHandler but unmatched, we do the same as if the tag was not supported so we don't lose the content.
On this branch, if we just unregister preformatted we don't lose the content it is pasted as a paragraph.

File: packages/blocks/src/api/raw-handling/utils.js [hunk 2]
```diff
@@ -37,18 +37,35 @@ export function getBlockContentSchema( transforms ) {
...
+			}
+			case 'attributes':
+			case 'require': {
+				return [ ...( objValue || [] ), ...( srcValue || [] ) ];
+			}
+			case 'isMatch': {
+				if ( ! objValue || ! srcValue ) {
+					return undefined;
+				}
+				return ( ...args ) => {
```
Comment (iseulde) [inline packages/blocks/src/api/raw-handling/utils.js]: Would it be possible to comment on this? It's hard to follow this if right away.

File: packages/blocks/src/api/raw-handling/utils.js [hunk 3]
```diff
@@ -37,18 +37,35 @@ export function getBlockContentSchema( transforms ) {
 				schema[ tag ].attributes.push( 'id' );
 			}
 		}
+		if ( isMatch ) {
```
Comment (iseulde) [inline packages/blocks/src/api/raw-handling/utils.js]: Same here. Would it be possible to comment.?

File: packages/blocks/src/api/raw-handling/utils.js [hunk 4]
```diff
@@ -154,7 +171,10 @@ function cleanNodeList( nodeList, doc, schema, inline ) {
 		const tag = node.nodeName.toLowerCase();
 
 		// It's a valid child.
-		if ( schema.hasOwnProperty( tag ) ) {
+		if (
+			schema.hasOwnProperty( tag ) &&
+			( ! schema[ tag ].isMatch || schema[ tag ].isMatch( node ) )
```
Comment (iseulde) [inline packages/blocks/src/api/raw-handling/utils.js]: Same, let's leave a comment?

**Review by ellatrix — APPROVED** [review]
Sounds good to me. Would love some more inline comments.

File: packages/blocks/src/api/raw-handling/utils.js [hunk 5]
```diff
@@ -37,18 +37,41 @@ export function getBlockContentSchema( transforms ) {
 				schema[ tag ].attributes.push( 'id' );
 			}
 		}
+
+		// If an isMatch function exists add it to each schema tag that it applies to.
+		if ( isMatch ) {
+			for ( const tag in schema ) {
+				schema[ tag ].isMatch = isMatch;
```
Comment (mcsf) [inline packages/blocks/src/api/raw-handling/utils.js]: Is this mutating the block type? Should it?

File: packages/blocks/src/api/raw-handling/utils.js (hunk 2 above)
Comment (mcsf) [reply packages/blocks/src/api/raw-handling/utils.js]: A description of the overall flow of the `mergeWith` call would help too.

File: packages/blocks/src/api/raw-handling/utils.js (hunk 5 above)
Comment (jorgefilipecosta) [reply packages/blocks/src/api/raw-handling/utils.js]: Hi @mcsf, the transforms passed to this function are always the result of getRawTransformations and this function copies the transforms, so we are mutating that copy and not the block type.
https://github.com/WordPress/gutenberg/blob/fix/media-text-i18n/packages/blocks/src/api/raw-handling/index.js#L57-L65
This function was already mutating the input, but I think given that it is a getter it is unexpected that this function mutates the input and I will open a follow-up PR that removes the mutations.
